[SPIR-V] Implement SPV_KHR_float_controls2

maarquitos14 opened the pull request:
Implementation of [SPV_KHR_float_controls2](https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_float_controls2.html) extension, and corresponding tests.

Some of the tests make use of `!spirv.ExecutionMode` LLVM named metadata. This is because some SPIR-V instructions don't have a direct equivalent in LLVM IR, so the SPIR-V Target uses different LLVM named metadata to convey the necessary information. Below, you will find an example from one of the newly added tests:
```
!spirv.ExecutionMode = !{!19, !20, !21, !22, !23, !24, !25, !26, !27}
!19 = !{ptr @k_float_controls_float, i32 6028, float poison, i32 131079}
!20 = !{ptr @k_float_controls_all, i32 6028, float poison, i32 131079}
!21 = !{ptr @k_float_controls_float, i32 31}
!22 = !{ptr @k_float_controls_all, i32 31}
!23 = !{ptr @k_float_controls_float, i32 4461, i32 32}
!24 = !{ptr @k_float_controls_all, i32 4461, i32 16}
!25 = !{ptr @k_float_controls_all, i32 4461, i32 32}
!26 = !{ptr @k_float_controls_all, i32 4461, i32 64}
!27 = !{ptr @k_float_controls_all, i32 4461, i32 128}
```
`!spirv.ExecutionMode` contains a list of metadata nodes, and each of them specifies the required operands for expressing a particular `OpExecutionMode` instruction in SPIR-V. For example, `!19 = !{ptr @k_float_controls_float, i32 6028, float poison, i32 131079}` will be lowered to `OpExecutionMode [[k_float_controls_float_ID]] FPFastMathDefault [[float_type_ID]] 131079`.

Commit:
Initial support for SPV_KHR_float_controls2.

Commit:
Fix test failures.

Commit:
Deprecations only apply when the extension is enabled.

Commit:
Add extension specific tests.

Commit:
Further work to apply deprecations only when extension is enabled.

Commit:
Remove leftover comment.

Commit:
Yet more work to apply deprecations only when extension is enabled.

Commit:
Renamings and bugfixes.

Commit:
Renamings, bugfixes and undo undesired changes.

Commit:
Undo undesired changes.

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Commit:
Fix after-merge bugs.

Commit:
Fix clang-format issues.

Commit:
Replace undef with poison for placeholders.

MrSidims commented on llvm/lib/Target/SPIRV/SPIRVUtils.cpp:
```diff
@@ -995,4 +1033,22 @@ unsigned getArrayComponentCount(const MachineRegisterInfo *MRI,
+    return 1;
+  case 64: // double
+    return 2;
+  case 128: // fp128
```
How does SPIR-V BE handle 128-bit fp type?

maarquitos14 replied:
We have an offline discussion and came to the conclusion that 128-bit fp type is not supported by SPIRV. In consequence, I will remove everything related to it in this PR, both in implementation and tests.

MrSidims commented on llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_float_controls2/exec_mode.ll:
```diff
@@ -0,0 +1,95 @@
+; CHECK-DAG: OpExecutionMode %[[#KERNEL_ALL]] FPFastMathDefault %[[#FLOAT_TYPE]] 262144 
+; CHECK-DAG: OpExecutionMode %[[#KERNEL_ALL]] FPFastMathDefault %[[#DOUBLE_TYPE]] 458752 
+; CHECK-DAG: OpExecutionMode %[[#KERNEL_ALL]] FPFastMathDefault %[[#FP128_TYPE]] 65543 
+!22 = !{ptr @k_float_controls_all, i32 6028, half poison, i32 131072}
```
Please add description in the PR, how !spirv.ExecutionMode works. Also https://github.com/search?q=repo%3Allvm%2Fllvm-project%20SPIRVRepresentationInLLVM&type=code has to be updated.

MrSidims commented on llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_float_controls2/exec_mode2.ll:
```diff
@@ -0,0 +1,58 @@
+!7 = !{i32 32, i32 36}
+!8 = !{i32 0, i32 0}
+!12 = !{i32 1, !"wchar_size", i32 4}
+!13 = !{!"clang version 8.0.1"}
```
guess we can remove unnecessary metadata

MrSidims commented on llvm/lib/Target/SPIRV/SPIRVAsmPrinter.cpp:
```diff
@@ -458,6 +459,7 @@ void SPIRVAsmPrinter::outputExecutionModeFromMDNode(
     unsigned ExpectMDOps, int64_t DefVal) {
   MCInst Inst;
   Inst.setOpcode(SPIRV::OpExecutionMode);
+
```
```suggestion
```

MrSidims commented on llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_float_controls2/exec_mode.ll:
```diff
@@ -0,0 +1,95 @@
+; CHECK-DAG: OpExecutionMode %[[#KERNEL_HALF]] FPFastMathDefault %[[#HALF_TYPE:]] 1 
+!17 = !{ptr @k_float_controls_half, i32 6028, half poison, i32 1}
+
+; CHECK-DAG: OpExecutionMode %[[#KERNEL_BFLOAT]] FPFastMathDefault %[[#BFLOAT_TYPE:]] 2 
```
This tingles my spider senses, bfloat type (aka encoding from SPV_KHR_bfloat extension) is not supported by the BE. I assume llvm bfloat is lowered just to TypeFloat 32 then. If it's the case, why [[#BFLOAT_TYPE:]] entry is different from [[#FLOAT_TYPE:]] entry in the test?

maarquitos14 replied:
`bfloat` is lowered to `OpTypeFloat 16`. `[[#BFLOAT_TYPE:]]` is different from `[[#HALF_TYPE:]]` in case it is eventually lowered to a different type.

MrSidims replied:
>OpTypeFloat 16

With [encoding](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#FP_Encoding) different from IEEE-754, right? Otherwise it's a bug in lowering (definitely not related to your patch, yet nice to know ahead of time).

maarquitos14 replied:
Actually, no. There is only a single `OpTypeFloat 16` without any encoding at all.

MrSidims reviewed (commented):
I'll take a closer look during next week. Some brief comments are listed below, the major one is a request to update SPIR-V friendly LLVM IR doc for the backend.

Commit:
Add SPIR-V Instructions Mapped to LLVM Metadata section.

Commit:
Address code review.

Keenuts reviewed (commented):
Started to review but I have a more global question first:

Seems like you do some stuff multiple times:
 - check the metadata on the module, and emit default decorations (`outputFPFastMathDefaultInfo`)
 - check metadata on all the rest, and emit decorations (later in the same `SPIRVAsmPrinter` function)
 - then merge flags you find from the OpDecorate instruction, checking for conflicts
 
Code seems a bit convoluted to me with parts that could definitely be refactored (ex the iteration on each type and MI construction for the decoration)

Looks like the metadata gathering, flag conflict validation and only then decoration emission could be split into distinct, independent pieces.

Another question: reading the float_control2 spec, seems like it could be enabled for Shaders since the capability is not gated by the Kernel one. But looks like the SPIRVAsmPrinter part is gated on `ST->isKernel`

Keenuts commented:
@s-perron

Commit:
Address code review feedback.

Commit:
Fix documentation build.

Commit:
Fix another error in documentation build.

maarquitos14 commented on llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_float_controls2/decoration.ll:
```diff
@@ -0,0 +1,92 @@
+  %remRes = frem arcp float %1,  %2
+  %negRes = fneg fast float %1
+  %oeqRes = fcmp nnan ninf oeq float %1,  %2
+  %oneRes = fcmp one float %1,  %2, !spirv.Decorations !3
```
Okay, so, to be clear, going forward we want to ignore every `spirv.Decoration` metadata that is related to FP flags, including:
- FPFastMathMode
- NoContraction

I will address this in this PR. Please, speak up if you're against this.

s-perron reviewed (commented):
I took a quick look. I'm okay with the general direction. I'd have to look into detail at the extensions to make sure the details are correct. I'll come back to that later. Do not hold up merging because of me.

Commit:
Ignore spirv.Decorations metadata related to FP flags.

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Keenuts reviewed (commented):
sorry thought I had published the comments.
Not very familiar with the float extensions & float behaviors in general, so couple observations. Will defer approval to others on the spec correctness, but OK to get this merged otherwise.

MrSidims commented on llvm/lib/Target/SPIRV/SPIRVAsmPrinter.cpp:
```diff
@@ -551,12 +568,53 @@ void SPIRVAsmPrinter::outputExecutionMode(const Module &M) {
+        for (const MachineInstr *MI :
+             MAI->getMSInstrs(SPIRV::MB_TypeConstVars)) {
+          // Skip if the instruction is not OpTypeFloat.
+          if (MI->getOpcode() != SPIRV::OpTypeFloat)
```
Somewhere should also be a OpTypeMatrix check, but I'm unsure, how SPIR-V backend works with shader matrices.

maarquitos14 replied:
Actually, this is the wording in the spec:
```
FPFastMathDefault
Set the default fast math flags for instructions not themselves decorated with FPFastMathMode. This only affects instructions operating on or resulting in a type that is Target Type or an OpTypeMatrix or OpTypeVector derived from it. **Target Type must be a scalar, floating-point type**. Fast-Math Mode must be the <id> of a [constant instruction](https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_float_controls2.html#ConstantInstruction) of 32-bit integer type containing a valid [FP Fast Math Mode](https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_float_controls2.html#FP_Fast_Math_Mode) bitmask. Fast-Math Mode must not be a specialization-constant instruction. May be applied at most once per Target Type to any execution mode.
```

So I think we should not. However, I will double-check if `handleMIFlagDecoration` is working properly for vector types.

maarquitos14 commented on llvm/lib/Target/SPIRV/SPIRVAsmPrinter.cpp:
```diff
@@ -603,6 +661,76 @@ void SPIRVAsmPrinter::outputAnnotations(const Module &M) {
+                             "SignedZeroInfNanPreserve but at least one of "
+                             "NotNaN/NotInf/NSZ is disabled.");
+
+        Flags |= SPIRV::FPFastMathMode::NotNaN | SPIRV::FPFastMathMode::NotInf |
+                 SPIRV::FPFastMathMode::NSZ;
```
Again, great catch. It doesn't, none of these flags should be enabled, as checked in the if just above it.

maarquitos14 replied:
Done in https://github.com/llvm/llvm-project/pull/146941/commits/f2aefbf502e6a48a5f8b251bdfbeb7ed8d412041.

Commit:
Address code review feedback.

maarquitos14 commented on llvm/test/CodeGen/SPIRV/extensions/SPV_KHR_float_controls2/decoration.ll:
```diff
@@ -0,0 +1,81 @@
+; Function Attrs: convergent mustprogress norecurse nounwind
+define weak_odr dso_local spir_kernel void @foo(float %1, float %2) {
+entry:
+  %addRes = fadd float %1,  %2
```
Done in https://github.com/llvm/llvm-project/pull/146941/commits/f2aefbf502e6a48a5f8b251bdfbeb7ed8d412041.

maarquitos14 commented on llvm/lib/Target/SPIRV/SPIRVUtils.cpp:
```diff
@@ -995,6 +1012,22 @@ unsigned getArrayComponentCount(const MachineRegisterInfo *MRI,
+  }
+  assert(false && "Unreachable code");
+  // This return is just to avoid compiler warnings.
+  return 0;
```
Done in https://github.com/llvm/llvm-project/pull/146941/commits/f2aefbf502e6a48a5f8b251bdfbeb7ed8d412041.

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Commit:
Fix clang-format.

Keenuts reviewed (commented):
Sorry for the delay, was OOO

maarquitos14 commented on llvm/lib/Target/SPIRV/SPIRVAsmPrinter.cpp:
```diff
@@ -606,6 +664,73 @@ void SPIRVAsmPrinter::outputAnnotations(const Module &M) {
+          MAI->getRegisterAlias(MI->getMF(), MI->getOperand(0).getReg());
+      Inst.addOperand(MCOperand::createReg(TypeReg));
+      unsigned Flags = FPFastMathDefaultInfo.FastMathFlags;
+      if (FPFastMathDefaultInfo.ContractionOff &&
+          (Flags & SPIRV::FPFastMathMode::AllowContract) &&
+          FPFastMathDefaultInfo.FPFastMathDefault)
```
You're right, it's illegal. However, because `ContractionOff` --and `SignedZeroInfNanPreserve`--, is deprecated when `SPV_KHR_float_controls2` is enabled, and the spec suggests to use `FPFastMathDefault` instead, what we're doing is simply replacing the former with the latter. It is explained in https://github.com/llvm/llvm-project/blob/848611acb1917b292c890990c7ca44b172aec8be/llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.h#L137C1-L145. Does that make sense?

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Commit:
Address reviewer feedback.

Commit:
Merge branch 'main' into maronas/float-controls2

Keenuts reviewed (commented):
Tried to figure out why you disabled spirv-val tests, and seems like there are things you can easily fix (OpExecutionModeId instead of OpExecutionMode) & fmin/fmax declarations for overloads, but one thing I'm not sure how to properly do is fix the bitmast emission. You need to getOrCreate an OpConstant with the flag value, but not sure what the cleanest option is. Do you have an issue to track those tasks?

Commit:
Use constant instruction for flags.

Commit:
Enable spirv-val and update tests accordingly.

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Commit:
Merge branch 'maronas/float-controls2-conflict' into maronas/float-controls2

Commit:
Rename new function to better represent what it does.

Commit:
Merge branch 'main' into maronas/float-controls2

maarquitos14 commented:
@Keenuts I have updated this PR to make `FPFastMathDefault` use constant instructions instead of immediates for the bitmask. A summary of the update: I added a new function in `SPIRVEmitIntrinsics` where I go through the `FPFastMathDefault/ContractionOff/SignedZeroInfNanPreserve` metadata, and create constants for the values that will be required for these execution modes. Then, in `SPIRVAsmPrinter`, I loop over the constants in the module and I build a hashtable where the key is the immediate value, and the value is the `MachineInstr` representing the constant. Then, when outputting the `OpExecutionMode` the hashtable is queried to get the constant. I think this was the least intrusive way of proceeding, but I'm open to suggestions :)

Commit:
Fix clang-format issues.

Commit:
Merge branch 'maronas/float-controls2-conflict' into maronas/float-controls2

maarquitos14 commented on llvm/lib/Target/SPIRV/SPIRVAsmPrinter.cpp:
```diff
@@ -552,12 +569,53 @@ void SPIRVAsmPrinter::outputExecutionMode(const Module &M) {
+          }
+
+          MCInst Inst;
+          Inst.setOpcode(SPIRV::OpExecutionMode);
```
Done.

Commit:
Remove unused variable.

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Commit:
Fix test failure.

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.h:
```diff
@@ -159,6 +186,13 @@ struct ModuleAnalysisInfo {
+  // first element is the smallest bit width, and the last element is the
+  // largest bit width, therefore, we will have {half, float, double} in
+  // the order of their bit widths.
+  DenseMap<const Function *, SmallVector<FPFastMathDefaultInfo, 3>>
```
nit: I'll be in favor or having `SmallVector<FPFastMathDefaultInfo, 3>` changed into:

```cpp
struct FPFastMathDefaultInfoVector : public SmallVector<FPFastMathDefaultInfo, 3> {
  static size_t computeFPFastMathDefaultInfoVecIndex(size_t BitWidth) {
     // the code from SPIRVUtils
```

This way you keep the index magic numbers close to the object using them.
Also, seems like there is a reimplementation of the FPFastMathDefaultInfo in the emit intrinsincs, meaning this couldn't be reused there. But once this code duplication removed, all `SmallVector<Info, 3>` bits can be replaced with this new class.

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:
```diff
@@ -262,6 +264,33 @@ class SPIRVEmitIntrinsics
   }
 };
 
+struct FPFastMathDefaultInfo {
```
This is an exact copy of the FPFastMathDefaultInfo struct in the ModuleAnalysis? Why not reuse it?

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVAsmPrinter.cpp:
```diff
@@ -665,15 +696,34 @@ void SPIRVAsmPrinter::outputAnnotations(const Module &M) {
+        continue;
+
+      if (OpCode == SPIRV::OpConstantI)
+        ConstMap[MI->getOperand(2).getImm()] = MI;
```
Ints could be signed, so here seems like you could end up with a `-1` being added as a max_uint or something like that no?

Commit:
Apply code review suggestions.

Commit:
Merge remote-tracking branch 'origin/main' into maronas/float-controls2

Commit:
Fix clang-format issue.

s-perron commented:
@Keenuts I on vacation for a while. You seem to have addressed his comment. You do not have to wait for him to merge.

Commit:
Merge branch 'main' into maronas/float-controls2

Commit:
Merge branch 'main' into maronas/float-controls2

Commit:
Address test failures caused by main branch merge.

Commit:
Merge remote-tracking branch 'origin/main' into HEAD

MrSidims reviewed (approved):
Test adjustment in https://github.com/llvm/llvm-project/pull/146941/commits/fe49f8b223cd23d65b01f12dbf9934a39d28b4bf is LGTM
